cupy/functional/piecewise.py

```diff
@@ -0,0 +1,58 @@
+import cupy
+
+from cupy import core
+
+_piecewise_krnl = core.ElementwiseKernel(
+    'U condlist, S funclist',
+    'T y',
+    ' if(condlist) y = funclist',
+    'piecewise_kernel'
+)
+
+
+def piecewise(x, condlist, funclist):
+    """Evaluate a piecewise-defined function.
+
+        Args:
+            x (cupy.ndarray): input domain
+            condlist (list of cupy.ndarray or bool scalars):
+                Each boolean array/ scalar corresponds to a function
+                in funclist. Length of funclist is equal to that
+                of condlist. If one extra function is given, it is used
+                as otherwise condition
+            funclist (cupy.ndarray or list): list of scalar functions.
+
+        Returns:
+            cupy.ndarray: the scalar values in funclist on portions of x
+                defined by condlist.
+
+        .. warning::
+
+            This function currently doesn't support callable functions,
+            args and kw parameters.
+
+        .. seealso:: :func:`numpy.piecewise`
+        """
+
+    if any(callable(item) for item in funclist):
+        raise NotImplementedError(
+            'Callable functions are not supported currently')
+    if cupy.isscalar(condlist) or (not isinstance(
+            condlist[0], (list, cupy.ndarray)) and x.ndim != 0):
+        condlist = [condlist]
+    condlist = cupy.array(condlist, dtype=bool)
+    condlen = len(condlist)
+    funclen = len(funclist)
+    if condlen == funclen:
+        y = cupy.zeros(shape=x.shape, dtype=x.dtype)
+    elif condlen + 1 == funclen:  # o.w
+        y = cupy.full(shape=x.shape, fill_value=funclist[-1], dtype=x.dtype)
+        funclist = funclist[:-1]
+        funclen -= 1
+    else:
+        raise ValueError('with {} condition(s), either {} or {} functions'
+                         ' are expected'.format(condlen, condlen, condlen + 1))
+    funclist = cupy.asarray(funclist)
```

Comment:
I will need this line if I am to use where in order to change funclist to x.dtype ?

Reply:
Please cast the type of `funclist` without using additional GPU memory space.

Reply:
Could you add tests for different data types of `x` and `funclist`?
The current implementation seems to raise different types of errors if `x` of integer type and `funclist` of complex type are given.

Reply:
> The current implementation seems to raise errors of different types if `x` of integer type and `funclist` of complex type are given.

Yes, this test fails but so is numpy's version

Reply:
So please fix to raise the error of the same type as `numpy.piecewise` raises.

Reply:
> Please cast the type of `funclist` without using additional GPU memory space.

I am allowing funclist to be either a list or cupy.array. So Should I only allow arrays and change the tests as well ? And for casting, use astype ?

Reply:
Testing for callable functions  or mixed function types is not possible when funclist is array

Reply:
Moving the loop ( that raises error if callable fns) is also dependent on when the cast is done or else testing for callable functions or mixed function types will fail. Also casting each function when passed as argument to kernel ( with no additional GPU memory space) fails because of mismatched data types when converting from input dtype to python type

Reply:
Why do you want to cast `funclist` to `x.dtype`?

Reply:
If not, tests will fail 

Consider

>> x = [True, False, True]
>> condlist = [[True, False, False],  [True, False, True],  [True, True, False]]
>> funclist = [1, 0, 3]
>> np.piecewise(x, condlist, funclist)
[True, True, False]

Reply:
I see. Please cast with `.astype(x.dtype, copy=False)` to avoid unnecessary memory allocation.

Reply:
It won't work with tuple. I had to use asarray() for tuple cases before astype()